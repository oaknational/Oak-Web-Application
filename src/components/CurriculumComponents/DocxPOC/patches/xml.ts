import { xml2js } from "xml-js";

export function xmlElementsToJson(xmlData: string) {
  return xml2js(xmlData, {
    compact: false,
    captureSpacesBetweenElements: false,
  });
}

export function xmlElementToJson(xmlData: string) {
  return xmlElementsToJson(xmlData).elements[0];
}
