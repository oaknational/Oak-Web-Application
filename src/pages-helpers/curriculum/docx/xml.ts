import { js2xml, xml2js } from "xml-js";
import type { Element } from "xml-js";

export function xmlRootToJson(xmlData: string) {
  return xml2js(xmlData, {
    compact: false,
    captureSpacesBetweenElements: false,
  });
}

/**
 * To support <https://prettier.io/docs/en/options#embedded-language-formatting> as well as assertion.
 *
 * NOTE: This must called `xml`
 *
 * @param strings template strings
 * @param values tamplate values
 * @returns xml string
 */
export function safeXml(
  strings: TemplateStringsArray,
  ...values: (string | number | string[])[]
) {
  let outXml = "";
  for (let i = 0; i < strings.length; i++) {
    if (values[i] !== undefined) {
      const valueRaw = values[i];
      const value = Array.isArray(valueRaw) ? valueRaw.join("") : valueRaw;
      outXml += `${strings[i]!}${value}`;
    } else {
      outXml += `${strings[i]!}`;
    }
  }

  // Just for assertion
  xmlRootToJson(outXml);

  return outXml;
}

export function xmlElementToJson(xmlData?: string) {
  if (!xmlData) return;
  const rootNode = xmlRootToJson(xmlData);
  if (rootNode.elements.length !== 1) {
    throw new Error("Expecting a single root node in XML");
  }
  return collapseFragments(rootNode as Element)?.elements?.[0];
}

export function createFragment(elements: Element[]): Element {
  return {
    type: "element",
    name: "XML_FRAGMENT",
    elements: elements,
  };
}

export function collapseFragments(root: Element) {
  const run = (orig: Element) => {
    let node = orig;
    const cloneIfRequired = () => {
      if (node === orig) {
        return {
          ...node,
          elements: node.elements ? [...node.elements] : undefined,
        };
      }
      return node;
    };
    if (node.elements) {
      for (let i = node.elements.length - 1; i >= 0; i--) {
        if (node.elements) {
          const child = node.elements[i]!;
          const out = run(child);

          if (Array.isArray(out)) {
            node = cloneIfRequired();
            node.elements!.splice(i, 1, ...out);
          } else if (out !== child) {
            node = cloneIfRequired();
            if (node.elements) {
              node.elements[i] = out;
            }
          }
        }
      }
    }

    if (node.name === "XML_FRAGMENT") {
      return node.elements ?? [];
    }
    return node;
  };

  const out = run(root);
  if (Array.isArray(out)) {
    return createFragment(out);
  }
  return out;
}

export function jsonXmlToXmlString(json: Element): string {
  const root = "name" in json ? { elements: [json] } : json;
  const output = js2xml(collapseFragments(root));
  return output;
}

export function cdata(input: string | number) {
  return `<![CDATA[${input}]]>`;
}

export function cdataJson(input: Element): Element {
  if (input.type !== "text") {
    throw new Error("Expecting text node");
  }
  return {
    type: "cdata",
    cdata: String(input.text),
  };
}
