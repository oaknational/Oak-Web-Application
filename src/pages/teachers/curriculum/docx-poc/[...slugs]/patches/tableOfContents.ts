import type { Element } from "xml-js";

import { textIncludes, textReplacer } from "./util";

export function tableOfContentsPatch() {
  return async (el: Element) => {
    if (el.type === "text" && textIncludes(el.text, "{{=TABLE_OF_CONTENTS}}")) {
      return {
        type: "text",
        text: textReplacer(el.text, "{{=TABLE_OF_CONTENTS}}", "TODO"),
      };
    }
    return el;
  };
}
