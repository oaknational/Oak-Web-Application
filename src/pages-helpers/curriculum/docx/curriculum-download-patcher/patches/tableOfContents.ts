import type { Element } from "xml-js";

import { cdataJson } from "../../xml";

import { textIncludes, textReplacer } from "./util";

export function tableOfContentsPatch() {
  return async (el: Element) => {
    if (el.type === "text" && textIncludes(el.text, "{{=TABLE_OF_CONTENTS}}")) {
      return cdataJson({
        type: "text",
        text: textReplacer(el.text, "{{=TABLE_OF_CONTENTS}}", "TODO"),
      });
    }
    return el;
  };
}
