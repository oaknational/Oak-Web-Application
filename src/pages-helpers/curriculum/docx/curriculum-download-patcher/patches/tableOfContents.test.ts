import { CombinedCurriculumData } from "..";
import { jsonXmlToXmlString } from "../../xml";

import { tableOfContentsPatch } from "./tableOfContents";

describe("tableOfContents", () => {
  const data = {
    units: [
      { year: "7" },
      { year: "9" },
      { year: "10" },
      { year: "8" },
      { year: "11" },
    ],
  } as CombinedCurriculumData;
  it("should modify node if present", async () => {
    const patcher = tableOfContentsPatch(data);

    const node = await patcher({
      type: "element",
      name: "w:p",
      elements: [
        {
          type: "element",
          name: "w:r",
          elements: [
            {
              type: "text",
              text: "testing {{=TABLE_OF_CONTENTS}} a node",
            },
          ],
        },
      ],
    });

    const xml = jsonXmlToXmlString(node);
    expect(xml).toContain("Year 7");
    expect(xml).toContain("Year 8");
    expect(xml).toContain("Year 9");
    expect(xml).toContain("Year 10");
    expect(xml).toContain("Year 11");
    expect(node).toMatchSnapshot();
  });

  it("should not modify invalid element", async () => {
    const patcher = await tableOfContentsPatch(data);

    const input = {
      type: "element",
      name: "w:p",
      elements: [
        {
          type: "element",
          name: "w:r",
          elements: [
            {
              type: "text",
              text: "testing {{=FOOBAR}} a node",
            },
          ],
        },
      ],
    };
    const node = await patcher(input);

    expect(node).toEqual(input);
  });
});
