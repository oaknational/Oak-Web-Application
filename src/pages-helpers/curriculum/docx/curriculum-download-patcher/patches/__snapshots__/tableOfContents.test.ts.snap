// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tableOfContents should modify node if present 1`] = `
{
  "elements": [
    {
      "elements": [
        {
          "elements": [
            {
              "elements": [
                {
                  "attributes": {
                    "w:val": "56",
                  },
                  "name": "w:sz",
                  "type": "element",
                },
                {
                  "attributes": {
                    "w:ascii": "Lexend Bold",
                    "w:cs": "Lexend Bold",
                    "w:eastAsia": "Lexend Bold",
                    "w:hAnsi": "Lexend Bold",
                  },
                  "name": "w:rFonts",
                  "type": "element",
                },
                {
                  "attributes": {
                    "w:val": "222222",
                  },
                  "name": "w:color",
                  "type": "element",
                },
              ],
              "name": "w:rPr",
              "type": "element",
            },
            {
              "attributes": {
                "xml:space": "preserve",
              },
              "elements": [
                {
                  "text": "Contents",
                  "type": "text",
                },
              ],
              "name": "w:t",
              "type": "element",
            },
          ],
          "name": "w:r",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "name": "w:p",
      "type": "element",
    },
    {
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "our_curriculum",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Our curriculum",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "what_are_threads",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Threads",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "curriculum_overview",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "undefined curriculum overview",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "year_7",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Year 7",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "year_8",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Year 8",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "year_9",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Year 9",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "year_10",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Year 10",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "year_11",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Year 11",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "attributes": {
            "w:anchor": "threads",
          },
          "elements": [
            {
              "elements": [
                {
                  "elements": [
                    {
                      "attributes": {
                        "w:val": "32",
                      },
                      "name": "w:sz",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:ascii": "Lexend Bold",
                        "w:cs": "Lexend Bold",
                        "w:eastAsia": "Lexend Bold",
                        "w:hAnsi": "Lexend Bold",
                      },
                      "name": "w:rFonts",
                      "type": "element",
                    },
                    {
                      "attributes": {
                        "w:val": "222222",
                      },
                      "name": "w:color",
                      "type": "element",
                    },
                  ],
                  "name": "w:rPr",
                  "type": "element",
                },
                {
                  "attributes": {
                    "xml:space": "preserve",
                  },
                  "elements": [
                    {
                      "cdata": "Threads appendix",
                      "type": "cdata",
                    },
                  ],
                  "name": "w:t",
                  "type": "element",
                },
              ],
              "name": "w:r",
              "type": "element",
            },
          ],
          "name": "w:hyperlink",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
    {
      "elements": [
        {
          "elements": [
            {
              "name": "w:pageBreakBefore",
              "type": "element",
            },
          ],
          "name": "w:pPr",
          "type": "element",
        },
      ],
      "name": "w:p",
      "type": "element",
    },
  ],
  "name": "$FRAGMENT$",
  "type": "element",
}
`;
